In [1]:
import pandas as pd

Loading Data and brief overview

In [4]:
data=pd.read_csv("matches.csv", index_col=0)

In [5]:
data.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1389 entries, 1 to 42
Data columns (total 27 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          1389 non-null   object 
 1   time          1389 non-null   object 
 2   comp          1389 non-null   object 
 3   round         1389 non-null   object 
 4   day           1389 non-null   object 
 5   venue         1389 non-null   object 
 6   result        1389 non-null   object 
 7   gf            1389 non-null   float64
 8   ga            1389 non-null   float64
 9   opponent      1389 non-null   object 
 10  xg            1389 non-null   float64
 11  xga           1389 non-null   float64
 12  poss          1389 non-null   float64
 13  attendance    693 non-null    float64
 14  captain       1389 non-null   object 
 15  formation     1389 non-null   object 
 16  referee       1389 non-null   object 
 17  match report  1389 non-null   object 
 18  notes         0 non-null      float

In [8]:
data.shape

(1389, 27)

In [10]:
data["team"].value_counts()

team
Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: count, dtype: int64

In [11]:
data[data["team"]=="Liverpool"]

date   time            comp         round  day venue result   gf  \
1   2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13  2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14  2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16  2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18  2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22  2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23  2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24  2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25  2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27  2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28  2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30  2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31  2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32  2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33  2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34  2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36  2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37  2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38  2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41  2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42  2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44  2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46  2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47  2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk  \
1   3.0     Leeds United  ...  Match Report    NaN  20.0   4.0  17.0  0.0   
2   0.0          Chelsea  ...  Match Report    NaN  17.0   5.0  17.7  1.0   
4   1.0          Arsenal  ...  Match Report    NaN  21.0   9.0  16.8  0.0   
6   7.0      Aston Villa  ...  Match Report    NaN  14.0   8.0  15.8  1.0   
7   2.0          Everton  ...  Match Report    NaN  22.0   8.0  15.0  1.0   
9   1.0    Sheffield Utd  ...  Match Report    NaN  17.0   5.0  18.2  1.0   
11  1.0         West Ham  ...  Match Report    NaN   8.0   2.0  18.6  1.0   
13  1.0  Manchester City  ...  Match Report    NaN   9.0   2.0  21.5  0.0   
14  0.0   Leicester City  ...  Match Report    NaN  24.0  12.0  11.9  0.0   
16  1.0         Brighton  ...  Match Report    NaN   6.0   2.0  20.9  0.0   
18  0.0           Wolves  ...  Match Report    NaN  11.0   6.0  16.6  1.0   
20  1.0           Fulham  ...  Match Report

There are 2 seasons worth of data, each team in the epl plays 38 matches and there are 20 teams. So we would expect there to be 1520 rows, however there are 1389. What has been concluded is that some rows (which represent matches) are missing, some teams have all 72 games, some 71 or 70. It makes sense for there to be 6 teams with 30+ matches because in the epl, 3 teams get promoted and 3 get relegated at the end of a season, so over two seasons there will be 6 teams with matches in the 30's (because we are only looking at matches in the epl, In addition this data was scraped while the 2022 season was still ongoing so we are missing some games for that whole season). But there is a 7th team, Liverpool, who was not relegated yet only has 38 matches, which simply means that we are missing a seasons worth of matches for Liverpool.

In [12]:
data["round"].value_counts()

round
Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: count, dtype: int64

Cleaning Data

In [15]:
data.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [16]:
data["date"]=pd.to_datetime(data["date"])

In [18]:
data.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

Feature engineering for baseline model

In [19]:
data["venue_bin"]=data["venue"].astype("category").cat.codes

In [26]:
data[["venue", "venue_bin",]]
## venue is an important feature/predictor (home ground advantage), however, given it is of the object dtype
## it can't be used by ML models so we create the venue_bin feature to numerically identify each venue.

venue  venue_bin
1   Away          0
2   Home          1
3   Home          1
4   Away          0
6   Home          1
..   ...        ...
38  Away          0
39  Home          1
40  Away          0
41  Away          0
42  Home          1

[1389 rows x 2 columns]

In [27]:
data["opp_bin"]=data["opponent"].astype("category").cat.codes
data[["opponent", "opp_bin"]]
## opponent is an important feature/predictor, however, given it is of the object dtype
## it can't be used by ML models so we create the opp_bin feature to numerically identify each team

opponent  opp_bin
1        Tottenham       18
2     Norwich City       15
3          Arsenal        0
4   Leicester City       10
6      Southampton       17
..             ...      ...
38       Tottenham       18
39  Crystal Palace        6
40         Everton        7
41   Newcastle Utd       14
42         Burnley        4

[1389 rows x 2 columns]

In [30]:
data["hour_of_day"]=data["time"].str.replace(":.+", "", regex=True).astype("int")
## some teams may perform better when they play during a particualr hour of day

In [36]:
data["day_bin"]=data["date"].dt.day_of_week
## similar idea as before, some teams may perform better, if they play on a particular day of the week

In [ ]:
data["target"]=(data["result"] == "W")
data["target"] = data["target"].astype("int")
data[["result", "target"]]
## Here we are setting our target column which is what our model will be predicting. 
## The model predicts wetcher a team will win or not, hence both Draw and Loss will be represented as zero

result  target
1       L       0
2       W       1
3       W       1
4       W       1
6       D       0
..    ...     ...
38      L       0
39      L       0
40      W       1
41      L       0
42      W       1

[1389 rows x 2 columns]

Initial model training and testing

In [50]:
data.shape

(1389, 32)

In [51]:
from sklearn.ensemble import RandomForestClassifier

In [52]:
model=RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)
##come back

In [ ]:
data

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   fk   pk  pkatt  season              team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022   Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022   Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022   Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022   Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022   Manchester City   
..  ...             ...  ...  ...  ...    ...     ...               ...   
38  4.0       Tottenham  ...  0.0  0.0    0.0    2021  Sheffield United   
39  2.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Sheffield United   
40  0.0         Everton  ...  0.0  0.0    0.0    2021  Sheffield United   
41  1.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Sheffield United   
42  0.0         Burnley  ...  0.0  0.0    0.0    2021  Sheffield United   

   venue_bin opp_bin hour_of_day  day_bin  target  
1          0      18          16        6       0  
2          1      15          15        5       1  
3          1       0          12        5       1  
4          0      10          15        5       1  
6          1      17          15        5       0  
..       ...     ...         ...      ...     ...  
38         0      18          19        6       0  
39         1       6          15        5       0  
40         0       7          19        6       1  
41         0      14          18        2       0  
42         1       4          16        6       1  

[1389 rows x 32 columns]

In [58]:
train=data[data["date"]<'2022-01-01']
test=data[data["date"]>'2022-01-01']

In [61]:
features=["venue_bin", "opp_bin", "hour_of_day", "day_bin"]

In [62]:
model.fit(train[features], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [63]:
results=model.predict(test[features])

Accuracy Inspection of initial model

In [68]:
from sklearn.metrics import accuracy_score

In [69]:
accuracy=accuracy_score(test["target"], results)
accuracy

0.6123188405797102

In [70]:
# The model attains 61% accuracy, reflecting moderate performance. 
# Reasonable predictive ability given the inherent randomness in match outcomes

In [71]:
resultant_frame=pd.DataFrame(dict(actual=test["target"], prediction=results))

In [ ]:
resultant_frame

actual  prediction
31       1           0
32       0           0
34       1           0
35       1           1
37       0           1
..     ...         ...
33       0           0
34       0           0
35       1           0
36       0           0
37       0           1

[276 rows x 2 columns]

In [74]:
cross_table=pd.crosstab(index=resultant_frame["actual"], columns=resultant_frame["prediction"])

In [75]:
cross_table

prediction    0   1
actual             
0           141  31
1            76  28

In [76]:
from sklearn.metrics import precision_score

In [78]:
precision_score(test["target"], results)
# The accuracy score indicates the percentage of correct predictions, 
# the cross table shows that win predictions were incorrect more often than correct. 
# Therefore, precision was used to evaluate how reliable win predictions are, 
# showing us that the model’s ability to correctly predict wins remains limited and can be improved.

0.4745762711864407

Further feature engineering

In [90]:
team_matches=data.groupby("team")
Arsenal_group=team_matches.get_group("Arsenal")
# group the records in the data frame by teams 
# so that we can calculate the rolling averages for different features.
# Then use these additional features for improved model performance

In [91]:
def rolling_avg(group, col, new_cols):
    group=group.sort_values("date")
    rolling_avgerages=group[col].rolling(3, closed='left').mean()
    group[new_cols]=rolling_avgerages
    group=group.dropna(subset=new_cols)
    return group

In [92]:
columns = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_columns= ["rol_gf", "rol_ga", "rol_sh", "rol_sot", "rol_dist", "rol_fk", "rol_pk", "rol_pkatt"]
# new_columns=[f"rol_{c}" for c in columns]

In [94]:
data_new=data.groupby("team").apply(lambda x: rolling_avg(x, columns, new_columns))


C:\Users\13327\AppData\Local\Temp\ipykernel_31904\2357531030.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  data_new=data.groupby("team").apply(lambda x: rolling_avg(x, columns, new_columns))


In [95]:
data_new

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_bin  target    rol_gf    rol_ga     rol_sh  \
team                                                                         
Arsenal                 6         6       1  2.000000  1.333333   7.666667   
                        7         5       0  1.666667  1.666667   5.333333   
                        9         6       0  1.000000  1.666667   7.000000   
                        11        6       1  0.666667  1.000000   9.666667   
                        13        6       0  0.333333  0.666667   9.666667   
...                             ...     ...       ...       ...        ...   
Wolverhampton Wanderers 32        6       1  1.333333  1.000000  12.333333   
                        33        4       0  1.666667  0.666667  12.333333   
                        34        5       1  2.333333  1.000000  13.000000   
                        35        4       0  1.666667  1.333333  13.000000   
                        36        6       0  1.333333  1.666667  10.000000   

                             rol_sot   rol_dist    rol_fk    rol_pk  rol_pkatt  
team                                                                            
Arsenal                 6   3.666667  14.733333  0.666667  0.000000   0.000000  
                        7   3.666667  15.766667  0.000000  0.000000   0.000000  
                        9   3.666667  16.733333  0.666667  0.000000   0.000000  
                        11  4.000000  16.033333  1.000000  0.000000   0.000000  
                        13  2.666667  18.033333  1.000000  0.333333   0.333333  
...                              ...        ...       ...       ...        ...  
Wolverhampton Wanderers 32  3.666667  19.300000  0.000000  0.000000   0.000000  
                        33  4.333333  19.600000  0.000000  0.000000   0.000000  
                        34  5.333333  19.833333  0.000000  0.000000   0.000000  
                        35  5.000000  18.533333  0.000000  0.000000   0

In [96]:
data_new=data_new.droplevel("team")

In [97]:
data_new.index=range(data_new.shape[0])

Retraining machine learning model

In [114]:
def predtion_testing(data, predictors):
    train=data[data["date"]<'2022-01-01']
    test=data[data["date"]>'2022-01-01']
    model.fit(train[predictors], train["target"])
    predictions=model.predict(test[predictors])
    resultant_frame2=pd.DataFrame(dict(actual=test["target"], predicted=predictions), index=test.index)
    precision=precision_score(test["target"], predictions)
    acc=accuracy_score(test["target"], predictions)
    return resultant_frame2, acc, precision


In [124]:
resultant_frame2, accuracy_result, precs=predtion_testing(data_new,features+new_columns)

In [125]:
accuracy_result


0.6666666666666666

In [126]:
precs
# Incorporating rolling averages from the previous three matches significantly improved model performance, 
# increasing precision from 47% to 62.5%. 
# Rolling features capture recent form and short-term momentum, 
# allowing the model to reflect current team performance more accurately than static season-level averages.

0.625